server/chat-plugins/help.ts

```diff
@@ -0,0 +1,473 @@
+	/** Stats - filter match and faq that was matched - done day by day. */
+	stats?: {[k: string]: DayStats};
+	/** Word pairs that have been marked as a match for a specific FAQ. */
+	pairs: {[k: string]: string[]};
+	/** Whether or not the filter is disabled. */
+	disabled?: boolean;
+	/** Queue of suggested regex. */
+	queue?: string[];
+}
+
+export class HelpAnswerer {
+	roomFaqs: AnyObject;
+	disabled?: boolean;
+	queue: string[];
+	data: PluginData;
+	constructor(data: PluginData) {
+		this.data = data;
+		this.roomFaqs = this.loadFaqs();
+		this.queue = data.queue ? data.queue : [];
+
+		FS(ROOMFAQ_FILE).onModify(() => {
+			// refresh on modifications to keep it up to date
+			this.roomFaqs = this.loadFaqs();
+		});
+	}
+	find(question: string, user?: User) {
+		const faqs: string[] = Object.keys((this.roomFaqs || '{}'))
+			.filter(item => item.length >= 6 && !this.roomFaqs[item].startsWith('>'));
+		for (const term of COMMON_TERMS) {
+			if (term.test(question)) return null;
+		}
+		for (const faq of faqs) {
+			const match = this.match(question, faq);
+			if (match) {
+				if (user) {
+					const timestamp = Chat.toTimestamp(new Date()).split(' ')[1];
+					const log = `${timestamp} |c| ${user.name}|${question}`;
+					this.log(log, faq, match.regex);
+				}
+				return this.roomFaqs[match.faq];
+			}
+		}
+		return null;
+	}
+	visualize(question: string, hideButton?: boolean, user?: User) {
+		const response = this.find(question, user);
+		if (response) {
+			let buf = '';
+			buf += Utils.html`<strong>You said:</strong> ${question}<br />`;
+			buf += `<strong>Our automated reply:</strong> ${Chat.formatText(response)}`;
+			if (!hideButton) {
+				buf += Utils.html`<hr /><button class="button" name="send" value="A: ${question}">`;
+				buf += `Send to the Help room if you weren't answered correctly or wanted to help </button>`;
+			}
+			return buf;
+		}
+		return null;
+	}
+	getFaqID(faq: string) {
+		faq = faq.trim();
+		if (!faq) return;
+		const alias: string = this.roomFaqs[faq];
+		if (!alias) return;
+		// ignore short aliases, they cause too many false positives
+		if (faq.length < MINIMUM_LENGTH || alias.length < MINIMUM_LENGTH) return;
+		if (alias.charAt(0) !== '>') return faq; // not an alias
+		return alias.replace('>', '');
+	}
+	stringRegex(str: string, raw?: boolean) {
+		str = str.split('=>')[0];
+		const args = str.split(',').map(item => item.trim());
+		return args.map(item => {
+			const split = item.split('&').map(string => {
+				// allow raw regex for admins and whitelisted users
+				if (raw) return string.trim();
+				// escape otherwise
+				return string.replace(/[\\^$.*+?()[\]{}]/g, '\\$&').trim();
+			});
+			return split.map(term => {
+				if (term.startsWith('!')) {
+					return `^(?!.*${term.slice(1)})`;
+				}
+				if (!term.trim()) return null;
+				return `(?=.*?(${term.trim()}))`;
+			}).filter(Boolean).join('');
+		}).filter(Boolean).join('');
+	}
+	validateRegex(regex: string) {
```

Comment:
does similar code already exist in other commands which accept regexes? (`/filter`? `/banwords`?). can it be consolidated?

Reply:
It does. I'll push that into a Chat function.